File: src/main/scala/unwrittenfun/minecraft/unwrittenblocks/items/ItemUpgrade.scala
package unwrittenfun.minecraft.unwrittenblocks.items

import java.util

import net.minecraft.client.renderer.texture.IconRegister
import net.minecraft.creativetab.CreativeTabs
import net.minecraft.entity.player.EntityPlayer
import net.minecraft.item.{ItemStack, Item}
import net.minecraft.util.{EnumChatFormatting, Icon}
import unwrittenfun.minecraft.unwrittenblocks._
import unwrittenfun.minecraft.unwrittenblocks.client.gui.creativetabs

/**
 * Project: UnwrittenBlocks
 * Author: UnwrittenFun 
 * Created: 11/08/2014.
 */
class ItemUpgrade(id: Int, keys: Array[String]) extends Item(id) {
  setUnlocalizedName("upgrade")
  setCreativeTab(creativetabs.ubTab)
  setHasSubtypes(true)

  var icons: Array[Icon] = new Array[Icon](keys.length)

  override def registerIcons(register: IconRegister) {
    for (keyId <- 0 to (keys.length - 1)) {
      icons(keyId) = register.registerIcon(TEXTURE_LOCATION + ":" + keys(keyId))
    }
  }

  override def getIconFromDamage(meta: Int): Icon = icons(meta)

  override def getSubItems(id: Int, tab: CreativeTabs, genericList: util.List[_]) {
    val itemList = genericList.asInstanceOf[(util.ArrayList[ItemStack])]

    for (meta <- 0 to (keys.length - 1)) {
      itemList add new ItemStack(id, 1, meta)
    }
  }

  override def getUnlocalizedName(stack: ItemStack): String = {
    keys(stack.getItemDamage)
  }

  override def addInformation(stack: ItemStack, player: EntityPlayer, genericList: util.List[_], par4: Boolean) {
    val list = genericList.asInstanceOf[(util.ArrayList[String])]

    stack.getItemDamage match {
      case 0 => list.add(EnumChatFormatting.GOLD + "Crafting Component")
      case 1 =>
        list.add(EnumChatFormatting.GOLD + "Increase Speed of Machines")
        list.add(EnumChatFormatting.GRAY + "Stacks up to 8")
      case _ =>
    }
  }
}

File: src/main/java/unwrittenfun/minecraft/unwrittenblocks/common/items/ItemUpgrade.java
package unwrittenfun.minecraft.unwrittenblocks.common.items;

import net.minecraft.client.renderer.texture.IIconRegister;
import net.minecraft.creativetab.CreativeTabs;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.item.Item;
import net.minecraft.item.ItemStack;
import net.minecraft.util.IIcon;
import unwrittenfun.minecraft.unwrittenblocks.common.ModInfo;

import java.util.Collections;
import java.util.List;

/**
 * Project: UnwrittenBlocks Author: UnwrittenFun Created: 04/11/2014.
 */
public class ItemUpgrade extends ItemUB {
  public String[] keys;
  public IIcon[]  icons;

  public ItemUpgrade(String mainKey, String[] keys) {
    super(mainKey);
    this.keys = keys;
    setHasSubtypes(true);
  }

  @Override public IIcon getIconFromDamage(int meta) {
    return icons[meta];
  }

  @Override public String getUnlocalizedName(ItemStack stack) {
    return "item." + keys[stack.getItemDamage()];
  }

  @Override public void addInformation(ItemStack stack, EntityPlayer player, List lines, boolean bool) {
    Collections.addAll(lines, getLinesFromLang("unwrittenblocks.text.info_" + keys[stack.getItemDamage()]));

    super.addInformation(stack, player, lines, bool);
  }

  @SuppressWarnings("unchecked") @Override public void getSubItems(Item upgrade, CreativeTabs creativeTabs,
                                                                   List items) {
    super.getSubItems(upgrade, creativeTabs, items);
    items.add(new ItemStack(upgrade, 1, 1));
  }

  @Override public void registerIcons(IIconRegister iconRegister) {
    icons = new IIcon[keys.length];
    for (int i = 0; i < icons.length; i++) {
      icons[i] = iconRegister.registerIcon(ModInfo.RESOURCE_LOCATION + ":" + keys[i]);
    }
  }
}
